Added support in Mavlink Ethernet channel parameters

garfieldG opened the pull request:
Mavlink Ethernet channel settings such as UDP port, remote port and broadcast mode now can be changed dynamically via parameters.

**Describe problem solved by this pull request**
-The start of the Mavlink Ethernet channel is static and not dynamic like the start of the serial channel.
-It allows us to dynamically choose Mavlink Ethernet channels settings via parameters.
-It allows us to use Broadcast on one Mavlink channel and Multicast on another.

**Describe your solution**
I added Ethernet related parameters to Mavlink.
Just like the serial parameters we had, the new parameters are times the number of Mavlink instances and are generated.
The new parameters will be generated only if the board supports Ethernet (by adding ETHERNET to its cmake).
Also, to choose Ethernet a new value was added to MAV_{I}_CONFIG and will be available only if the board supports Ethernet (otherwise the new value won't be generated as an option).

**Additional context**
I tested it on a costume board and it works fine.
Previously I [suggested](https://github.com/PX4/Firmware/pull/10124) adding parameters such as IP address, Mac address.
In my code, I also added parameters to choose remote multicast address and another line in the generated Mavlink command that supports multicast, however, the added line only works if the multicast address is known from the parameters. I will gladly add it if you want.

dagar commented:
Slightly lazy question, how does this extend multiple instances? For example on SITL (or a linux board) where we have multiple instances on different ports. Does specifying the board has ETHERNET give you one additional (parameter configurable and autostarted) Mavlink instance?

dagar commented:
@dinomani this might be of interest for v5x.

garfieldG commented:
@dagar Hi, I'm not sure that I understand your question, but I'll try to answer.
It doesn't give any extra instances, you still have the same number of possible Mavlink instances, but you are now able to configure any of these Mavlink instances to use ethernet instead of serial via parameters.
When you specify your board has ETHERNET :
- ethernet is added as a value that you can select :
![image](https://user-images.githubusercontent.com/27761531/77524875-c785c480-6e90-11ea-8b9d-1cb1f3f9f526.png)
you can choose ethernet on one instance or more. 
not like the other values, ethernet can be chosen multiple times.
 -new parameters are added :
![image](https://user-images.githubusercontent.com/27761531/77525307-817d3080-6e91-11ea-9f03-7345b54cff17.png)
![image](https://user-images.githubusercontent.com/27761531/77525345-8d68f280-6e91-11ea-9733-7fa0cb6ed609.png)
![image](https://user-images.githubusercontent.com/27761531/77525372-9a85e180-6e91-11ea-9095-7ea9da14048b.png)

note that the extra parameters and value are to all instances and not just to instance 0.

one more thing, this change is not only for Mavlink, if in the future there will be any parameters that you would like to generate only for boards that support ethernet then you can always do that by using the ethernet keyword inside the yaml file.

garfieldG commented:
@dagar @dinomani 
Hi, 
Is there anything new regarding this PR? Would you like me to add or change something?

BazookaJoe1900 commented:
@dagar can you take a look on that?

dagar commented:
Sorry, I don't recall where we got to with this one.

Having a configurable ports per mavlink instance feels wrong, but we can live with it. What if I wanted to have more than 1 mavlink instance (eg SITL)?

garfieldG commented:
> Having configurable ports per mavlink instance feels wrong
I'm not sure I agree with you about that. I did it because we encountered a problem in which we needed more than one mavlink instance and each one of the instances needed to be working with a different port. It needed to be configurable because our need changed from time to time so we wanted it to be parameters configurable with no need to change something in the code each time just for a different port.
> What if I wanted to have more than 1 mavlink instance
No problem here. Just like it was before you have MAV_{i}_CONFIG. You add mavlink instances just like you did it before, the only change is that now you can select that your mavlink instances will work with ethernet instead of serial. you can have more than one ethernet instance, because unlike the regular serial options you can select in MAV_{i}_CONFIG the ethernet option can be selected multiple times, if you chose it once you can choose it again as many times that you want (and as many max instances you have of course). Also, the ethernet option will only appear as a possible option only if your board supports ethernet (PX4_ETHERNET).

BazookaJoe1900 commented:
@dagar ping....

hamishwillee commented on src/modules/mavlink/module.yaml:
```diff
@@ -95,3 +104,45 @@ parameters:
+            default: [14556, 0, 0]
+            ethernet:
+
+        MAV_${i}_REMOT_PORT:
```
REMOT or REMOTE?

garfieldG replied:
@hamishwillee REMOT... REMOTE will bring us to 17 chars and 16 is the limit

hamishwillee replied:
Thanks @garfieldG I guess I should have counted. FWIW I probably would have done REMOTE_PRT, although I guess a matter of taste

BazookaJoe1900 replied:
I agree with @hamishwillee REMOTE_PRT looks better

garfieldG replied:
@hamishwillee Changed it. Also changed UDP_PORT to UDP_PRT to be similar

hamishwillee replied:
Thanks @garfieldG .I wouldn't have suggested you do that (even for consistency) but it is completely fine :-).

hamishwillee commented:
If this goes in it may need documentation. Probably and update to https://docs.px4.io/master/en/peripherals/mavlink_peripherals.html

LorenzMeier commented:
Could you rebase? Sorry that this didn't get pushed forward earlier.

garfieldG commented:
@LorenzMeier Thanks for the update. Just rebased it now.

LorenzMeier commented:
Thanks!

garfieldG commented:
@LorenzMeier Hi, I rebased again and added the rc.mavlink_override script as discussed in the dev call.

Just as a side note, we use parameters to decide the mac address and IP address. In the past, it didn't seem to get too much appreciation, but I would gladly pr it again now if there's any interest.

mrpollo commented:
@davids5 can you make sure that this falls into your plans for mavlink/ethernet networking support, and its aligned?

davids5 commented:
> @davids5 can you make sure that this falls into your plans for mavlink/ethernet networking support, and its aligned?

This is just the configuration and is not setting the nodes's IP address. So good to go from my perspective.

@garfieldG - Would you please rebase this, so we can test it and merge it.

garfieldG commented:
@davids5 rebased it, thanks!

davids5 commented:
@garfieldG - Thank you, I have set it to auto merge. But if CI fails ping me.

garfieldG commented:
@davids5 The CI failed

mrpollo commented:
I suspect the failures have nothing to do with this PR, I restarted the jobs just to be sure, but the errors were mostly segmentation faults and timeouts.

I will monitor the results of CI and merge if they are indeed false positives.

bkueng commented on Tools/serial/generate_config.py:
```diff
@@ -211,6 +224,80 @@ def parse_yaml_parameters_config(yaml_config):
         param_group = parameters_section.get('group', None)
         for param_name in definitions:
             param = definitions[param_name]
+            if 'ethernet' in param:
```
Can you avoid the code duplication here, and instead pass `has_ethernet_support`?

bkueng commented on Tools/serial/generate_config.py:
```diff
@@ -339,7 +428,8 @@ def parse_yaml_parameters_config(yaml_config):
             'param_group': port_config['group'],
-            'description_extended': port_config.get('description_extended', '')
+            'description_extended': port_config.get('description_extended', ''),
+			'ethernet_config': serial_command.get('ethernet', 'none')
```
`'ethernet_config': serial_command.get('ethernet', None)`

bkueng commented on validation/module_schema.yaml:
```diff
@@ -206,5 +210,9 @@ parameters:
                             # [0, N-1]
                             # Default: 0
                             type: integer
+                        ethernet:
```
Can you change this to `requires_ethernet: type: boolean`?
